File: Delete nodes having greater value on right - GFG/delete-nodes-having-greater-value-on-right.java
//{ Driver Code Starts
import java.util.*;

class Node
{

    int data;
    Node next;

    Node(int d)
    {
        data = d;
        next = null;
    }
}


public class LinkedList
{
    
 

    /* Function to print linked list */
   public static void print(Node root)
    {
        Node temp = root;
        while(temp!=null)
        {
            System.out.print(temp.data + " ");
            temp=temp.next;
        }
    }


    /* Driver program to test above functions */
    public static void main(String args[])
    {

        /* Constructed Linked List is 1.2.3.4.5.6.
         7.8.8.9.null */
        int value;
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t > 0 && sc.hasNextInt() )
        {
            int n = sc.nextInt();
            
            //int n=Integer.parseInt(br.readLine());
            int a1 = sc.nextInt();
            Node head = new Node(a1);
            Node tail = head;

            for (int i = 1; i < n && sc.hasNextInt() ; i++)
            {
                int a = sc.nextInt();
                tail.next = new Node(a);
                tail = tail.next;
            }


            Node result = new Solution().compute(head);
            print(result);
            System.out.println();
            t--;
        }
    }
}
// } Driver Code Ends


/*
class Node {
   int data;
   Node next;

  Node(int data) {
      this.data = data;
  }
}
*/
class Solution
{
    Node reverse(Node head){
        Node ptr = head, pre = null;
        while(ptr != null){
            Node nex = ptr.next;
            ptr.next = pre;
            pre = ptr;
            ptr = nex;
        }
        return pre;
    }
    Node compute(Node head)
    {
        // your code here
        if(head == null){
            return head;
        }
        Node re = reverse(head);
        Node ans = new Node(re.data);
        Node tr = ans;
        int mx = re.data;
        re = re.next;
        while(re != null){
            if(re.data >= mx){
                tr.next = new Node(re.data);
                mx = re.data;
                tr = tr.next;
            }
            re = re.next;
        }
        return reverse(ans);
    }
}
  

File: Delete nodes having greater value on right - GFG/delete-nodes-having-greater-value-on-right.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>

using namespace std;

struct Node
{
    int data;
    Node* next;
    
    Node(int x){
        data = x;
        next = NULL;
    }
};


void print(Node *root)
{
    Node *temp = root;
    while(temp!=NULL)
    {
        cout<<temp->data<<" ";
        temp=temp->next;
    }
}



// } Driver Code Ends
/*

The structure of linked list is the following

struct Node
{
    int data;
    Node* next;
    
    Node(int x){
        data = x;
        next = NULL;
    }
};
*/
class Solution
{
    public:
    Node* reverse(Node* head){
        Node* ptr = head, *pre = NULL;
        while(ptr != NULL){
            Node* nex = ptr->next;
            ptr->next = pre;
            pre = ptr;
            ptr = nex;
        }
        return pre;
    }
    Node *compute(Node *head)
    {
        // your code goes here
        if(head == NULL){
            return head;
        }
        Node* re = reverse(head);
        Node* ans = new Node(re->data);
        Node* tr = ans;
        int mx = re->data;
        re = re->next;
        while(re != NULL){
            if(re->data >= mx){
                tr->next = new Node(re->data);
                mx = re->data;
                tr = tr->next;
            }
            re = re->next;
        }
        return reverse(ans);
    }
    
};
   


//{ Driver Code Starts.

int main()
{
    int T;
	cin>>T;

	while(T--)
	{
		int K;
		cin>>K;
		struct Node *head = NULL;
        struct Node *temp = head;

		for(int i=0;i<K;i++){
		    int data;
		    cin>>data;
			if(head==NULL)
			    head=temp=new Node(data);
			else
			{
				temp->next = new Node(data);
				temp = temp->next;
			}
		}
        Solution ob;
        Node *result = ob.compute(head);
        print(result);
        cout<<endl;
    }
}

// } Driver Code Ends